### DISCUSSION TITLE
How to use upsert while PK is generated on application side?

### DISCUSSION THREAD

**Discussion by bill-o-more** [mikro-orm/mikro-orm#4532] [Q&A] [answered]
I'm following [this example](https://mikro-orm.io/docs/defining-entities#using-custom-base-entity) from the docs to generate the default values in the CustomBaseEntity:
```typescript
import { v4 } from 'uuid';

export abstract class CustomBaseEntity {

  @PrimaryKey()
  uuid = v4();

  @Property()
  createdAt = new Date();

  @Property({ onUpdate: () => new Date() })
  updatedAt = new Date();

}
```

Then I have an app init process which seeds the DB from the external content system, so I upsert the entities I get from there:
```typescript
const dbField = await em.upsert(MyEntity, {
            label, [the rest of the fields] // <-- label is unique
        });
// tried em.getRepository(MyEntity).upsert(...) with the same result
```

This is the query I get:
```sql
insert into "my_entity" ("label", [the rest of the fields]) values ('myLabel', [the rest of the values]) on conflict ("label") do update set "otherField" = excluded."other_field", "anotherField" = excluded."another_field" returning "id", "created_at", "updated_at"
```

which gives me the following error:
```
 null value in column "id" of relation "my_entity" violates not-null constraint
```

What do I miss?

**Comment by B4nan** [comment, marked as answer]
You need to have a database default for this, e.g. in postgres you can do:

```ts
@PrimaryKey({ type: 'uuid', defaultRaw: 'gen_random_uuid()' )
uuid = v4();
```

Alternatively, you could use the query builder.

**Reply by bill-o-more** [reply]
imagined something like that, but didn't know if it's gonna bite me in the ass at some point, since it's not in the docs. So this thread is now the docs :) thanks!

**Reply by bill-o-more** [reply]
though it looks like I can't upsert with M2M relation:
```typescript
MyEntityRepo.upsert({
  [my fields]
  m2mRelation: [anotherEntity],
});
```
tries to insert NULL instead of the newly upserted entity ID into the pivot table:
```sql
insert into "my_entity_another_entity" ("my_entity_id", "another_entity_id") values (NULL, '25352c52-88e4-4d79-9f33-fc9276ca5ad1') returning "my_entity_id", "another_entity_id"
```
which, of course, throws

UPDATE: Looks like I need to create it first, and then add a relation
UPDATE 2: yeah, this indeed works:
```typescript
const myEntity = MyEntityRepo.upsert({
  [my fields]
});
myEntity.m2mRelation = [anotherEntity];
```

UPDATE 3: but now it's not a part of the upsert, so it tries to add it again on the second run and fails because such relation already exists 🤷‍♂️ I can check if it exists before inserting, but maybe I miss some upsert'ish way of doing that?

**Reply by B4nan** [reply]
Upsert wont handle any relations, it is an abstraction over the `insert on conflict` queries that is identity map aware, thats all.

**Reply by bill-o-more** [reply]
love how it tried tho :)
